azure-pipelines.yml

```diff
@@ -0,0 +1,92 @@
+variables:
+  appDir: 'fakebook/'
+
+trigger:
+  branches:
+    include:
+    - main
+
+  paths:
+    include:
+    - azure-pipelines.yml
+    - 'fakebook'
+
+pr:
+  branches:
+    include:
+    - '*'
+
+  paths:
+    include:
+    - azure-pipelines.yml
+    - 'fakebook'
+
+pool:
+  vmImage: 'ubuntu-latest'
+
+stages:
+  - stage: build
+    jobs:
+    - job: build
+      variables:
+        imageName: 'fakebookng'
+      
+      steps:
+      - task: Docker@2
+        displayName: Build Docker Image
+        inputs:
+          repository: $(imageName)
+          command: 'build'
+          Dockerfile: '$(appdir)Dockerfile'
+  
+  - stage: SonarCloud
+    dependsOn: []
+    jobs:
+    - job: build
+      steps:
+      # using node version 15.x to enforce npm 7.x
+      - task: NodeTool@0
+        inputs:
+          versionSpec: '15.x'
+        displayName: 'Install Node.js'
+
+      - script: npm ci
+        displayName: npm install
+        workingDirectory: $(appDir)
+
+      - script: npx ng build --prod
+        displayName: ng build
+        workingDirectory: $(appDir)
+      
+      - script: npx ng lint
+        displayName: ng lint
+        workingDirectory: $(appDir)
+      - task: SonarCloudPrepare@1
+        condition: always()
+        displayName: 'sonarcloud prepare'
+        inputs:
+          SonarCloud: 'SonarCloud Token'
+          organization: '2011-fakebook-project3'
+          scannerMode: 'CLI'
+          configMode: 'manual'
+          cliProjectKey: '2011-fakebook-project3_ng'
+          cliProjectName: 'ng'
+          cliSources: '$(appDir)/src'
+          extraProperties: sonar.javascript.lcov.reportPaths="$(Agent.TempDirectory)/coverage/$(appDir)lcov.info"
+      
+      - script: npx ng test --browsers ChromeHeadless --watch false --code-coverage
+        displayName: ng test
+        workingDirectory: $(appDir)
+
+      - script: cat $(Agent.TempDirectory)/coverage/$(appDir)lcov.info 
```

Comment:
it's not in `Agent.TempDirectory`, it's just right there in the same sources directory everything else is in, `appDir`.

you should really drop the `/` from the end of the value of `appDir` though (and use `$(appDir)/` throughout instead), it's pretty brittle as-is

Reply:
Changed and fixed through the file